test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJmap.java

```diff
@@ -0,0 +1,183 @@
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+package org.openj9.test.attachAPI;
+
+import static org.openj9.test.attachAPI.TestConstants.TARGET_VM_CLASS;
+import static org.openj9.test.util.StringUtilities.searchSubstring;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+import static org.testng.AssertJUnit.assertTrue;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+import org.openj9.test.util.HelloWorld;
+import org.testng.annotations.AfterTest;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.BeforeSuite;
+import org.testng.annotations.Test;
+
+import com.ibm.tools.attach.attacher.OpenJ9VirtualMachine;
+import com.sun.tools.attach.AttachNotSupportedException;
+import com.sun.tools.attach.VirtualMachine;
+
+/**
+ * Test the jmap utility.
+ */
+@Test(groups = { "level.extended" })
+public class TestJmap extends AttachApiTest {
+
+	private static final String JMAP_COMMAND = "jmap"; //$NON-NLS-1$
+	private static String myId;
+	
+	/* test objects */
+	private ArrayList<HelloWorld> hws;
+	private int[][] primitiveArray;
+	private long[] primitiveVector;
+	private String[][] objectArray;
+	private Integer[] objectVector;
+
+	@Test
+	public void testHisto() throws IOException {
+		List<String> jmapOutput = runCommand(Arrays.asList(myId, "-histo")); //$NON-NLS-1$
+		logJmapOutput(jmapOutput);
+		checkHeapStats(jmapOutput);
+	}
+
+	@Test
+	public void testHistoLive() throws IOException {
+		List<String> jmapOutput = runCommand(Arrays.asList(myId, "-histo:live")); //$NON-NLS-1$
+		logJmapOutput(jmapOutput);
+		checkHeapStats(jmapOutput);
+	}
+
+	@Test
+	public void testHeapHisto() throws AttachNotSupportedException, IOException {
+		TargetManager tgtMgr = new TargetManager(TARGET_VM_CLASS, null);
+		assertTrue(AttachApiTest.CHILD_PROCESS_DID_NOT_LAUNCH, tgtMgr.syncWithTarget());
+		VirtualMachine vm = VirtualMachine.attach(tgtMgr.targetId);
+		assertTrue(vm instanceof OpenJ9VirtualMachine, "Wrong attach API VirtualMachine"); //$NON-NLS-1$
+		OpenJ9VirtualMachine myVm = (OpenJ9VirtualMachine) vm;
+		InputStream result = myVm.heapHisto();
+		checkForProperties(result);
+		result = myVm.heapHisto("-live"); //$NON-NLS-1$
+		checkForProperties(result);
+		result = myVm.heapHisto("-all"); //$NON-NLS-1$
```

Comment:
Should the `-all` test perhaps be run first so it has a chance of being different than `-live` (which implies a GC)?

Reply:
Good idea! Done.